### PR TITLE
fix: ignore stale observer map array notifications

### LINKED ISSUE #7579: fix: prevent stale observerMap array notifications after replacement in fast-html
# 🐛 Bug Report

ObserverMap-managed arrays can keep stale array observer subscriptions after an array branch is replaced by `deepMerge`. If external code retains a reference to the old root array, or to a nested array inside the replaced branch, later mutations to that stale array can still notify the owning FAST element's root property even though the current data no longer contains that array.

This is a follow-up hardening issue for #7559.

## 💻 Repro or Code Sample

Conceptual repro:

```ts
const oldOrders = element.user.orders;
const oldItems = oldOrders[0].items;

deepMerge(element.user, {
    orders: [
        {
            id: 103,
            items: [{ name: "Replacement", metrics: { views: 1 } }],
        },
    ],
});

oldOrders.push({ id: 104, items: [] });
oldItems.push({ name: "Stale", metrics: { views: 2 } });
```

The stale `oldOrders` and `oldItems` references should no longer be able to drive notifications or renders for the current `element.user` tree.

Also cover a shared-array case:

```ts
const shared = [{ name: "shared" }];
ownerA.data.items = shared;
ownerB.data.items = shared;

deepMerge(ownerA.data, { items: [{ name: "replacement" }] });
shared.push({ name: "still owned by B" });
```

Mutating `shared` should still notify `ownerB`, but not `ownerA`.

## 🤔 Expected Behavior

Array observer registrations should be scoped to the owner/root property/schema context that installed them.

When an observerMap-managed root or nested array is no longer reachable from a target's current root data, mutations to that stale array should not call `Observable.notify(target, rootProperty)` for that target. If the same array instance is still reachable from another owner, that other owner should continue to receive notifications.

## 😯 Current Behavior

The current implementation tracks observed arrays with a `WeakSet`, which only models whether an array has been observed at least once. It does not distinguish between multiple owner/root-property/schema contexts for shared arrays, and stale subscriptions on replaced arrays may still be live.

## 💁 Possible Solution

Replace `observedArraysMap: WeakSet<object>` with per-array registrations, for example `WeakMap<object, ArrayObserverRegistration[]>`.

Each registration should store the subscriber object passed to `Observable.getNotifier(data).subscribe(...)`. `subscribe()` returns `void`; it does not return a subscriber/disposable.

Track enough context to dedupe only the same observation context while allowing shared arrays to be observed by multiple owners: `target`, `rootProperty`, item `schema`, `rootSchema`, and `subscriber`.

Guard array subscriber work by checking that the changed array is still reachable from `target[rootProperty]` for that registration before proxying added items or notifying the root property. A root-level check such as `target[rootProperty] === subject` is not sufficient because nested arrays are valid subjects too.

If explicit teardown is added, apply it consistently from both replacement paths: the setter created by `defineObservableProperty` and the `assignProxy` `set` trap. Teardown should remove only registrations for the relevant owner context and must not unsubscribe registrations still needed by another owner or by a shared branch that remains reachable.

Suggested acceptance coverage:

- A Playwright test that holds a reference to the original root array, calls `deepMerge` to replace it, mutates the stale reference, and asserts that no parent notification/render is triggered.
- A Playwright test that holds a reference to a nested array inside the replaced branch, mutates it after replacement, and asserts that no parent notification/render is triggered.
- A Playwright test where the same array instance is observed by two owners; after one owner replaces its property, mutations through the shared array still notify the owner that continues to reference it.
- Coverage for both `deepMerge` replacement and direct property/proxy assignment replacement paths.

## 🔦 Context

PR #7559 intentionally changed observerMap `deepMerge` behavior to replace arrays instead of mutating observed arrays in place, avoiding synchronous reentrant array notification work. This follow-up closes remaining edge cases around stale retained references, nested arrays, and shared array instances.

Searched existing issues for overlapping observerMap/deepMerge/stale array subscription reports and did not find a duplicate.

## 🌍 Your Environment

* OS & Device: Not OS-specific
* Browser: Not browser-specific
* Version: `@microsoft/fast-html` after #7559

### PR DESCRIPTION
**Pull request by pradeep-ramola**
# Pull Request

## 📖 Description

Fixes stale observer-map array notifications after arrays are replaced.

Previously, observer-map tracked observed arrays with a single `WeakSet`, so the first owner/root context for an array could keep receiving notifications even after that array was no longer reachable from the current data. This changes array observation to use owner-specific registrations and guards notifications by checking whether the changed array is still reachable from the registered owner’s current root property.

This is a fix and is not a breaking change.

### 🎫 Issues

Fixes #7579

## 👩‍💻 Reviewer Notes

The main area to review is the new array registration/reachability logic in `observer-map-utilities.ts`.

A useful smoke test is the observer-map deep-merge fixture. It covers stale root arrays, stale nested arrays, direct assignment replacement, proxy assignment replacement, and shared arrays where only the owner that still references the array should be notified.

## 📑 Test Plan

- `npm run build:tsc -w @microsoft/fast-element`
- `npx biome check packages/fast-element/src/declarative/observer-map-utilities.ts packages/fast-element/test/declarative/fixtures/extensions/observer-map-deep-merge/main.ts packages/fast-element/test/declarative/fixtures/extensions/observer-map-deep-merge/observer-map-deep-merge.spec.ts change/@microsoft-fast-element-ab7cde20-cff1-47a9-8e95-5895dd961321.json`
- `npx beachball check --verbose`
- `npx playwright test --config=packages/fast-element/playwright.declarative.config.ts --project=chromium fixtures/extensions/observer-map-deep-merge/observer-map-deep-merge.spec.ts`

### General

- [x] I have included a change request file using `$ npm run change`
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/microsoft/fast/blob/main/CONTRIBUTING.md) documentation and followed the [standards](https://github.com/microsoft/fast/blob/main/CODE_OF_CONDUCT.md#our-standards) for this project.

### Agents

- [x] I have linked to an existing issue in this project that this change addresses
- [x] I have read the skills
- [x] I have read the DESIGN.md file(s) in packages relevent to my changes
- [ ] I have updated the DESIGN.md file(s) in packages relevent to my changes

## ⏭ Next Steps

None.

### CODE REVIEW

File: packages/fast-element/src/declarative/observer-map-utilities.ts [hunk 1]
```diff
@@ -156,6 +170,153 @@ function findArrayItemDef(schema: JSONSchema | JSONSchemaDefinition): string | n
...
+
+function observeArray(
+    data: any[],
+    reachableArray: any,
+    schema: JSONSchema | JSONSchemaDefinition,
+    rootSchema: JSONSchema,
+    target: any,
+    rootProperty: string,
+): void {
+    let registrations = observedArraysMap.get(data);
```
Comment (janechu) [inline packages/fast-element/src/declarative/observer-map-utilities.ts]: registrations are never cleaned up, a subscriber may have an old notifier attached, can you ensure these get cleaned up?

File: packages/fast-element/test/declarative/fixtures/extensions/observer-map-deep-merge/main.ts [hunk 2]
```diff
@@ -362,6 +366,225 @@ class DeepMergeTestElement extends FASTElement {
...
+        notifier.unsubscribe(subscriber, "users");
+
+        return {
+            notifications,
+            currentOrderCount: this.users[0].orders.length,
+            staleOrderCount: oldOrders.length,
+        };
+    }
+
+    public async mutateStaleNestedItemsAfterDeepMerge() {
```
Comment (janechu) [inline packages/fast-element/test/declarative/fixtures/extensions/observer-map-deep-merge/main.ts]: This test does not actually fail on the issue, consider this change:
```ts
public async mutateStaleNestedItemsAfterSecondDeepMerge() {
    deepMerge(this.users[0], {
        orders: [this.createOrder(901, "First replacement")],
    });
    await Updates.next();

    const oldItems = this.users[0].orders[0].items;

    deepMerge(this.users[0], {
        orders: [this.createOrder(902, "Second replacement")],
    });
    await Updates.next();

    let notifications = 0;
    const notifier = Observable.getNotifier(this);
    const subscriber = {
        handleChange() {
            notifications++;
        },
    };

    notifier.subscribe(subscriber, "users");

    oldItems.push(this.createProduct(9000, "Stale item"));
    await Updates.next();

    notifier.unsubscribe(subscriber, "users");

    return {
        notifications,
        currentItemCount: this.users[0].orders[0].items.length,
        staleItemCount: oldItems.length,
    };
}
```

Pair this with the following test in the spec file:
```ts
test("should ignore stale nested array mutations after replacing arrays with deepMerge", async ({
    page,
}) => {
    const hydrationCompleted = page.waitForFunction(
        () => (window as any).hydrationCompleted === true,
    );
    await page.goto("/fixtures/extensions/observer-map-deep-merge/");
    await hydrationCompleted;

    const result = await page
        .locator("deep-merge-test-element")
        .evaluate((element: any) =>
            element.mutateStaleNestedItemsAfterSecondDeepMerge(),
        );

    expect(result.notifications).toBe(0);
    expect(result.currentItemCount).toBe(1);
    expect(result.staleItemCount).toBe(2);
});
```

This pattern fails without the fix because  oldItems.push(...)  incorrectly notifies the element’s users root property after the second replacement.

**Review by janechu — CHANGES_REQUESTED** [review]
Thanks for the contribution @pradeep-ramola! I've added a few comments but overall this looks like a good fix.

File: packages/fast-element/test/declarative/fixtures/extensions/observer-map-deep-merge/main.ts (hunk 2 above)
Comment (pradeep-ramola) [reply packages/fast-element/test/declarative/fixtures/extensions/observer-map-deep-merge/main.ts]: Thanks, updated in 8937b496e.

- stale array registrations are now removed and their array notifier subscriber is unsubscribed
- the nested stale-array regression now uses the second deepMerge repro case from your comment

The focused Playwright fixture test, tsc build, Biome check, and Beachball check all pass locally.

**Comment by janechu** [comment]
Thanks for updating this. The stale-registration cleanup and the revised nested stale test address my earlier concerns.

I found one remaining edge case worth covering/fixing: shared primitive-array aliases within the same owner/root can lose notifications. Primitive arrays are wrapped in separate proxies but registered by the same raw array/context, and `observeArray()` updates the existing registration's `reachableArray`. If two current paths share the same raw primitive array, then one path is replaced, mutating the still-current other proxy can be treated as stale and skip `Observable.notify()`.

Concrete scenario I reproduced against this branch:

```ts
const sharedTags = ["shared"];
const itemA = element.users[0].orders[0].items[0];
const itemB = element.users[0].orders[0].items[1];

itemA.tags = sharedTags;
itemB.tags = sharedTags;
await Updates.next();

itemB.tags = ["replacement"];
await Updates.next();

itemA.tags.push("still owned by A");
await Updates.next();
```

`itemA.tags` remains current and includes the new value, but the root `users` notification count is `0` after the push. I think the fix is to track all reachable primitive-array wrappers/aliases for a registration, or otherwise distinguish registrations by wrapper/path enough that replacing one alias does not invalidate another still-reachable alias.

**Comment by pradeep-ramola** [comment]
> Thanks for updating this. The stale-registration cleanup and the revised nested stale test address my earlier concerns.
> 
> I found one remaining edge case worth covering/fixing: shared primitive-array aliases within the same owner/root can lose notifications. Primitive arrays are wrapped in separate proxies but registered by the same raw array/context, and `observeArray()` updates the existing registration's `reachableArray`. If two current paths share the same raw primitive array, then one path is replaced, mutating the still-current other proxy can be treated as stale and skip `Observable.notify()`.
> 
> Concrete scenario I reproduced against this branch:
> 
> ```ts
> const sharedTags = ["shared"];
> const itemA = element.users[0].orders[0].items[0];
> const itemB = element.users[0].orders[0].items[1];
> 
> itemA.tags = sharedTags;
> itemB.tags = sharedTags;
> await Updates.next();
> 
> itemB.tags = ["replacement"];
> await Updates.next();
> 
> itemA.tags.push("still owned by A");
> await Updates.next();
> ```
> 
> `itemA.tags` remains current and includes the new value, but the root `users` notification count is `0` after the push. I think the fix is to track all reachable primitive-array wrappers/aliases for a registration, or otherwise distinguish registrations by wrapper/path enough that replacing one alias does not invalidate another still-reachable alias.

Thanks, updated in d9db2f5fb.

This now tracks all reachable primitive-array aliases and notifies the owning root when a nested shared alias is still current after another alias is replaced. I also added the regression case from your comment for shared `tags` arrays.

**Comment by janechu** [comment]
Thanks, the shared raw primitive-array alias case looks addressed now. I found one closely related remaining case: assigning an already-observed primitive-array proxy as another observed property can double-wrap the proxy and produce invalid splice records.

Repro against the latest branch:

```ts
const itemA = element.users[0].orders[0].items[0];
const itemB = element.users[0].orders[0].items[1];

itemA.tags = ["shared"];
await Updates.next();

itemB.tags = itemA.tags;
await Updates.next();

itemB.tags.push("via B");
await Updates.next();
```

`itemA.tags` and `itemB.tags` both contain `['shared', 'via B']`, but the array observer emitted a splice record with `addedCount: 3` for the single push, and the fixture rendered extra empty `.tag` entries for `itemB`.

Suggested fix: mark observer-map primitive-array proxies with an internal symbol/flag (similar in spirit to `$isProxy` for objects), and in `assignObservablesToArray()` return/reuse the existing primitive-array proxy instead of creating a proxy around a proxy. The registration should still call `observeArray()`/`addReachableArray()` for the existing proxy so alias reachability is tracked, but the underlying array should have only one primitive-array proxy layer participating in splice generation.

**Review by janechu — DISMISSED** [review]

**Review by janechu — APPROVED** [review]
